program-runtime/src/compute_budget.rs

```diff
@@ -276,6 +280,99 @@ impl ComputeBudget {
             .map(|fee_type| PrioritizationFeeDetails::new(fee_type, self.compute_unit_limit))
             .unwrap_or_default())
     }
+
+    /// Calculate fee for `SanitizedMessage`
+    pub fn calculate_fee(
```

Comment:
Not feeling `compute_budget` is the right place for `calculate_fee()`. Initially think `calculate_fee()` can sit nicely in `sdk` right next to `FeeStructure`, but that would create circular dependency between `sdk` and `program-runtime`, since `calculate_fee()` needs `solana-compute-budget-program`. I see two options: 
1. create a new crate `solana-transaction-fee` that depends on `solana-compute-budget-program` and `solana-sdk`; The pro is we will have one place to do all fee related logic; the cons are another crate, and doesn't really address the root of dependency (which is why compute-budget-program is needed when calculate transaction base fee).
2. move `compute_budget.process_instructions()` all the way up to early stage of transaction lifetime, saying during transaction sanitization, so sanitized_transaction will have compute_budget results. Pros is `calculation_fee()` will only need fee_structure and sanitized_transaction itself, which makes better sense; and we may call compute_budget only once per transaction (a big plus). cons: a lot of change, may not even backpoortable. 

wdyt @apfitzge @t-nelson

Reply:
I'm not sure I favor either of these options. 2 seems like too big a change and will probably break a bunch of stuff. 1 seems fine in terms of another crate, but agree it doesn't really solve the dependence on `ComputeBudget`.

What would you think of having an additional layer of separation. `ComputeBudgetInstruction` already lives inside `sdk`. We could:

1. `fn extract_compute_budget_instructions(message: SanitizedMessage) -> ComputeBudgetInstructions`
2. `calculate_fees` uses internally, `calculate_fees_with_compute_budget_instructions`

No circular dependence. We can re-use the extracted CBI for both ComputeBudget creation, and fee calculation.

Maybe I'm off target, lmk.

Reply:
Trying to understand it better. Let me know if it aligns with your suggestion?

We add `extract_compute_budget_instructions()` and `calculate_fees_with_compute_budget_instructions()` to `sdk/src/fee.rs` (or maybe `sdk/src/compute_budget.rs`?), and reimplement `calculate_fees()` using these two. 
Move `calculate_fees()` to `sdk/src/fee.rs`.

Wondering if it'll cause any code maintenance overheads?
1. `extract_compute_budget_instructions()` would need to be in sync with `ComputeBudgetInstruction`, in case we add any new instructions in the future.
2. `calculate_fees_with_compute_budget_instructions()` would require replicating some of the `ComputeBudget::process_instructions()` code to `sdk`.

What are your thoughts on moving the `process_instructions()` code to `sdk`? Specifically the part that iterates over the instructions, and computes the values for the following:
```
        let mut num_non_compute_budget_instructions: usize = 0;
        let mut updated_compute_unit_limit = None;
        let mut requested_heap_size = None;
        let mut prioritization_fee = None;
        let mut updated_loaded_accounts_data_size_limit = None;
```

This new function can be used by the existing `process_instruction()` and `calculate_fee()`.  The existing `process_instruction()` will also update the state of `ComputeBudget` object.